File: python/be/api/steps/earn_steps.py
"""Steps for be-cex-earn.feature. Mirror of node/be/api/steps/earn.steps.js.
Staking locks a principal that accrues a fixed APR over explicit simulated
seconds and is redeemed for principal + reward.
"""

from __future__ import annotations

import re

from pytest_bdd import given, parsers, then, when

from be.api.venues.minicex import ApiUnreachable, MiniCex

cex = MiniCex()


def coin(n):
    return round(float(n) * 1e8)


def acct(qa, alias):
    return qa.accounts[alias]


def act(qa, fn):
    if qa.source_error:
        return
    try:
        fn()
    except ApiUnreachable as err:
        qa.source_error = str(err)


@then("inserting a stake with a non-positive principal fails a CHECK")
def stake_principal_check(qa):
    qa.tmp.execute("INSERT INTO accounts (id, handle, created_at) VALUES (1,'x',0)")
    try:
        qa.tmp.execute("INSERT INTO stakes (account_id, asset, principal, apr_bps, status, created_at) VALUES (1,'BTC',0,500,'active',0)")
        qa.check("principal>0 enforced", False, "no error thrown")
    except Exception as err:  # noqa: BLE001
        qa.check("principal>0 enforced", re.search(r"CHECK|constraint", str(err), re.I) is not None, str(err))


def do_stake(qa, alias, amt, asset, apr):
    a = acct(qa, alias)
    qa.stake_token = a["token"]
    qa.api = cex.stake(a["token"], {"asset": asset, "amount": coin(amt), "apr_bps": apr})
    if qa.api["status"] in (200, 201):
        qa.current_stake = qa.api["body"]


@given(parsers.parse('"{alias}" stakes {amt:g} {asset} at {apr:d} bps'))
@when(parsers.parse('"{alias}" stakes {amt:g} {asset} at {apr:d} bps'))
def stakes(qa, alias, amt, asset, apr):
    act(qa, lambda: do_stake(qa, alias, amt, asset, apr))


@given(parsers.parse("the stake accrues {seconds:d} seconds"))
@when(parsers.parse("the stake accrues {seconds:d} seconds"))
def stake_accrues(qa, seconds):
    def go():
        qa.api = cex.stake_accrue(qa.stake_token, qa.current_stake["id"], seconds)
        qa.current_stake = qa.api["body"]
    act(qa, go)


@given("the stake is redeemed")
@when("the stake is redeemed")
def stake_redeemed(qa):
    def go():
        qa.api = cex.stake_redeem(qa.stake_token, qa.current_stake["id"])
        qa.current_stake = qa.api["body"]
    act(qa, go)


@when("the stake is redeemed again")
def stake_redeemed_again(qa):
    act(qa, lambda: setattr(qa, "api", cex.stake_redeem(qa.stake_token, qa.current_stake["id"])))


@when(parsers.parse('"{alias}" tries to withdraw {amt:g} {asset}'))
def tries_withdraw(qa, alias, amt, asset):
    act(qa, lambda: setattr(qa, "api", cex.withdraw(acct(qa, alias)["token"], asset, coin(amt))))


def _assert(qa, description, cond, detail):
    if qa.source_error:
        qa.unobservable(description, qa.source_error)
        return
    qa.check(description, bool(cond), detail)


@then(parsers.parse("the stake accrued is {amt:g} {a2}"))
def stake_accrued(qa, amt, a2):
    _assert(qa, "stake accrued " + str(amt), qa.current_stake and qa.current_stake["accrued"] == coin(amt), f"got {qa.current_stake['accrued']} want {coin(amt)}" if qa.current_stake else "no stake")


@then("the second redemption was an idempotent replay")
def redeem_replay(qa):
    _assert(qa, "redeem idempotent replay", qa.api and qa.api["body"] and qa.api["body"].get("idempotent_replay") is True, str(qa.api["body"])[:80] if qa.api else "no response")

File: node/be/api/steps/earn.steps.js
'use strict';

const { Given, When, Then } = require('@cucumber/cucumber');
const { MiniCex, ApiUnreachable } = require('../venues/minicex');

/**
 * Steps for features/be-cex-earn.feature. Staking locks a principal that
 * accrues a fixed APR over explicit simulated seconds and is redeemed for
 * principal + reward. The account/deposit and ledger steps are shared; here are
 * the stake actions and stake-specific assertions.
 */

const cex = new MiniCex();
const coin = (n) => Math.round(Number(n) * 1e8);
const acctOf = (world, alias) => world.accounts[alias];

async function act(world, fn) {
  if (world.sourceError) return;
  try { await fn(); } catch (err) { if (err instanceof ApiUnreachable) { world.sourceError = err.message; return; } throw err; }
}
function expectThrow(fn, needle) {
  try { fn(); return { passed: false, detail: 'no error thrown' }; }
  catch (err) { const ok = new RegExp(needle, 'i').test(err.message); return { passed: ok, detail: ok ? 'rejected' : 'wrong error: ' + err.message }; }
}

Then('inserting a stake with a non-positive principal fails a CHECK', function () {
  this.tmp.exec("INSERT INTO accounts (id, handle, created_at) VALUES (1, 'x', 0)");
  const r = expectThrow(() => this.tmp.exec("INSERT INTO stakes (account_id, asset, principal, apr_bps, status, created_at) VALUES (1,'BTC',0,500,'active',0)"), 'CHECK|constraint');
  this.check('principal>0 enforced', r.passed, r.detail);
});

async function doStake(world, alias, amt, asset, apr) {
  const a = acctOf(world, alias);
  world.stakeToken = a.token;
  world.api = await cex.stake(a.token, { asset, amount: coin(amt), apr_bps: apr });
  if (world.api.status === 201 || world.api.status === 200) world.currentStake = world.api.body;
}
// One definition per phrase (Cucumber shares Given/When/Then); serves both the
// "And ..." setup lines and the "When ..." action lines.
When('{string} stakes {float} {word} at {int} bps', async function (alias, amt, asset, apr) { await act(this, () => doStake(this, alias, amt, asset, apr)); });
When('the stake accrues {int} seconds', async function (seconds) { await act(this, async () => { this.api = await cex.stakeAccrue(this.stakeToken, this.currentStake.id, seconds); this.currentStake = this.api.body; }); });
When('the stake is redeemed', async function () { await act(this, async () => { this.api = await cex.stakeRedeem(this.stakeToken, this.currentStake.id); this.currentStake = this.api.body; }); });
When('the stake is redeemed again', async function () { await act(this, async () => { this.api = await cex.stakeRedeem(this.stakeToken, this.currentStake.id); }); });

When('{string} tries to withdraw {float} {word}', async function (alias, amt, asset) {
  await act(this, async () => { this.api = await cex.withdraw(acctOf(this, alias).token, asset, coin(amt)); });
});

Then('the stake accrued is {float} {word}', function (amt, _a) {
  if (this.sourceError) { this.unobservable('stake accrued', this.sourceError); return; }
  this.check('stake accrued ' + amt, this.currentStake && this.currentStake.accrued === coin(amt), this.currentStake ? 'got ' + this.currentStake.accrued + ' want ' + coin(amt) : 'no stake');
});
Then('the second redemption was an idempotent replay', function () {
  if (this.sourceError) { this.unobservable('redeem replay', this.sourceError); return; }
  this.check('redeem idempotent replay', this.api && this.api.body && this.api.body.idempotent_replay === true, this.api ? JSON.stringify(this.api.body).slice(0, 80) : 'no response');
});
